lattice_controller_engine/lib/lattice_controller_engine/observed/lattice.ex

```diff
@@ -0,0 +1,474 @@
+    }
+
+    actor_instances =
+      if Enum.member?(instances, instance) do
+        instances
+      else
+        [instance | instances]
+      end
+
+    %Lattice{
+      l
+      | actors: Map.put(l.actors, pk, actor_instances),
+        instance_tracking:
+          Map.put(l.instance_tracking, instance.id, timestamp_from_iso8601(stamp))
+    }
+  end
+
+  defp put_provider_instance(
+         l = %Lattice{},
+         source_host,
+         pk,
+         link_name,
+         contract_id,
+         instance_id,
+         spec,
+         stamp
+       ) do
+    provider = Map.get(l.providers, {pk, link_name}, Provider.new(pk, link_name, contract_id))
+    instances = provider.instances
+
+    instance = %Instance{
+      id: instance_id,
+      host_id: source_host,
+      spec_id: spec
+    }
+
+    prov_instances =
+      if Enum.member?(instances, instance) do
+        instances
+      else
+        [instance | instances]
+      end
+
+    provider = %{provider | instances: prov_instances}
+
+    %Lattice{
+      l
+      | providers: Map.put(l.providers, {pk, link_name}, provider),
+        instance_tracking:
+          Map.put(l.instance_tracking, instance.id, timestamp_from_iso8601(stamp))
+    }
+  end
+
+  defp remove_provider_instance(l, source_host, pk, link_name, instance_id, spec) do
+    provider = Map.get(l.providers, {pk, link_name})
+
+    instance = %Instance{
+      id: instance_id,
+      host_id: source_host,
+      spec_id: spec
+    }
+
+    if provider == nil do
+      l
+    else
+      provider_instances =
+        Enum.reject(provider.instances, fn tgt_instance -> instance == tgt_instance end)
+
+      provider = %Provider{provider | instances: provider_instances}
+
+      %Lattice{
+        l
+        | providers: Map.put(l.providers, {pk, link_name}, provider),
+          instance_tracking: Map.delete(l.instance_tracking, instance.id)
+      }
+    end
+  end
+
+  defp record_host(l = %Lattice{}, source_host, labels, stamp) do
+    host = %Host{
+      id: source_host,
+      labels: labels,
+      last_seen: timestamp_from_iso8601(stamp)
+    }
+
+    %Lattice{l | hosts: Map.put(l.hosts, source_host, host)}
+  end
+
+  @spec timestamp_from_iso8601(binary) :: DateTime.t()
+  def timestamp_from_iso8601(stamp) when is_binary(stamp) do
+    case DateTime.from_iso8601(stamp) do
+      {:ok, datetime, 0} -> datetime
+      _ -> DateTime.utc_now()
+    end
+  end
+
+  @spec running_instances(
+          LatticeControllerEngine.Observed.Lattice.t(),
+          nil | String.t(),
+          String.t()
+        ) :: [%{id: String.t(), instance_id: String.t(), host_id: String.t()}]
+  def running_instances(%Lattice{} = l, pk, spec_id) when is_binary(pk) do
+    if String.starts_with?(pk, "M") do
+      actors_in_appspec(l, spec_id)
+      |> Enum.map(fn %{actor_id: pk, instance_id: iid, host_id: hid} ->
+        %{id: pk, instance_id: iid, host_id: hid}
+      end)
+    else
+      providers_in_appspec(l, spec_id)
```

Comment:
Because this is the fallback, is there a possible case where the `spec_id` is invalid and not corresponding to an actor or a provider? if that's the case, you may need a `provider_in_appspec` definition that is a fallthrough for an invalid spec_id

Reply:
noted, good catch.